### PR TITLE
feat: add sensor.community air quality monitoring (#20)

### LINKED ISSUE #20: Air quality from luftdaten / sensor.community
- https://data.sensor.community/airrohr/v1/filter/area=42.605,23.188,42.83,23.528 gives data within the Sofia bounding box (`SOFIA_BOUNDS` in source)
- This data includes all kinds of sensors. We are mostly interested in P1 and P2 sensors
- We don't want to push notifications based on a single sensor's flucatuations. Sensors are usually low-cost and often produce wrong data
- We need to implement rolling window
- We may want to put the sensors into "clusters" or grid so that a single sensor does not produce notifications, but an average of sensors does
- We want to display this as AQI but need to label it specifically as standard AQI is based on 24-hour monitoring window
- Wе want to push whenever we get AQI > 150 but delay for smaller increases and keep it based on rolling window. This way  we don't spam users because of a small spike but we also warn them when their attention is needed.

### PR DESCRIPTION
**Pull request by vbuch**
## What

Add real-time air quality monitoring using [sensor.community](https://sensor.community/) PM2.5/PM10 sensors in Sofia.

Closes #20

### Core feature
- **Crawler** (`ingest/crawlers/sensor-community/`) — fetches readings from sensor.community API, aggregates by grid cell, calculates AQI, and produces source documents with precomputed GeoJSON
- **AQI calculation** (`ingest/lib/air-quality/aqi.ts`) — PM2.5/PM10 breakpoint interpolation following EU AQI scale (Good → Very Poor)
- **Grid aggregation** (`ingest/lib/air-quality/grid.ts`) — groups sensors into geographic grid cells for spatial averaging
- **Outlier filtering** (`ingest/lib/air-quality/outlier-filter.ts`) — IQR-based statistical filtering to remove faulty sensor readings
- **Sensor response parsing** (`ingest/lib/air-quality/parse-sensor-response.ts`) — validates and normalizes raw API responses

### Storage (GCS)
- **`ReadingsStore`** (`ingest/lib/air-quality/readings-store.ts`) — dual-backend abstraction:
  - `GcsReadingsBackend` — production (one JSON blob per locality in GCS bucket)
  - `LocalReadingsBackend` — development (local filesystem)
- Rolling 24h retention window with append + dedup + prune on each fetch
- Replaces the original Firestore-based storage (~90% cost reduction by eliminating per-document write/index charges)

### Fetch scheduler
- **`air-quality-fetch.ts`** — Cloud Run job that fetches sensor data every 15 minutes and stores via `ReadingsStore`

### Infrastructure
- Terraform: Cloud Run v2 job, Cloud Scheduler (`*/15 * * * *`), GCS bucket variable
- Cloud Workflows: crawler step added to `all.yaml` and `emergent.yaml`

### Tests
- 63 unit tests across AQI, grid, outlier filter, sensor parsing, and readings store modules

### Documentation
- `docs/features/air-quality-monitoring.md` — feature overview, data flow, AQI scale
- `docs/features/air-quality-storage.md` — GCS vs local FS architecture

### CODE REVIEW

**Comment by vbuch** [comment]
## Team Review — PR #307 Status Report

Four-perspective review (End User, Business Analyst, Developer, QA) with all findings independently verified.

---

### BLOCKERS (must fix before merge)

| # | Issue | Verified | Impact |
|---|-------|----------|--------|
| **B1** | **`sensor-community` not registered in `shared/src/sources.ts`** | ✅ file has 17 sources, none is `sensor-community` | Feature invisible to API/UI; messages rejected |
| **B2** | **`geoJson` stored as string** — `index.ts` line 278 uses `JSON.stringify(cell.geoJson)` | ✅ confirmed | Grid polygons won't render on the map |
| **B3** | **Duplicate `severity: "ERROR"`** in `all.yaml` lines 325-326 | ✅ confirmed | YAML parse error, workflow deployment fails |
| **B4** | **No GCS bucket resource** in Terraform | ✅ no `google_storage_bucket` in `main.tf` | Bucket must be created manually or pipeline fails |
| **B5** | **No GCS IAM permissions** for service account | ✅ no `storage.*` role bindings exist | 403 Forbidden when writing readings |

---

### WARNINGS (should fix)

| # | Issue | Source |
|---|-------|--------|
| W1 | `ParsedReading` / `SensorReading` — duplicate types, consolidate into one | Developer |
| W2 | Unnecessary `new Date(r.timestamp)` when `.timestamp` is already a `Date` | Developer |
| W3 | No integration tests for `air-quality-fetch.ts` or crawler `crawl()` function | QA |
| W4 | No tests for API failure, timeout, zero readings, all-outliers edge cases | QA |
| W5 | No mock fixture with Polygon geometry for front-end testing | End User |

---

### SUGGESTIONS (nice to have)

| # | Suggestion | Source |
|---|------------|--------|
| S1 | Add crawler unit tests (`computeHourlyAverages`, alert decision logic, dedup URL) | Developer, QA |
| S2 | Document nighttime evaluation gap (7 AM–10:30 PM only) | End User |
| S3 | Add GCS lifecycle policy (auto-delete >2 days) | BA |
| S4 | Document grid-based approach for users | End User |
| S5 | Add explicit test for timezone "Z" suffix handling | Developer |

---

### Overall Assessment

**Architecture & code quality: strong.** Clean separation, good test coverage on individual modules (63 tests), NowCast AQI and IQR filtering well-implemented.

**Production readiness: blocked.** Five verified issues prevent deployment — all straightforward fixes (source registration, remove `JSON.stringify`, fix YAML dupe, add Terraform bucket + IAM). No architectural rework needed.

File: docs/features/air-quality-monitoring.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+Each alert is created as a source document with the grid cell's polygon geometry, the `air-quality` category, and a timespan covering the current evaluation half.
+
+## AQI Scale
+
+The system uses the US EPA Air Quality Index with Bulgarian labels:
+
+| AQI Range | Level |
+|-----------|-------|
+| 0–50 | Добро (Good) |
+| 51–100 | Умерено (Moderate) |
```
[the remark this answers is not in the document]
Comment (vbuch) [reply docs/features/air-quality-monitoring.md]: So what do we really calculate? It makes sense to use EU as we are targetting EU users. But are we?

> The US EPA AQI and EU Air Quality Index (mainly CAQI) differ in scale, thresholds, and health focus. The US uses a 0–500 scale focused on regulatory standards (NAAQS) and public health, while the EU typically uses a 0–100+ scale (or similar bands) focused on local exposure, often aligning closer to [WHO guidelines](https://www.who.int/publications/i/item/9789240034228).

**Comment by vbuch** [comment]
## Team Review — Post-Revision (EU EAQI + PR Comment Fixes)

### End User Perspective (Product)

| # | Type | Finding |
|---|------|---------|
| 1 | ✅ OK | Alert messages use clear Bulgarian text with EU EAQI labels ("Добро" through "Изключително лошо") — consistent with informal register per AGENTS.md |
| 2 | ✅ OK | Markdown-formatted alerts include PM2.5/PM10 values and sensor count for transparency |
| 3 | 💡 SUGGESTION | Consider whether `deepLinkUrl: ""` (empty string) should point to an external AQ dashboard or sensor.community map — users currently have no way to "learn more" about an alert |
| 4 | ✅ OK | Grid cell polygons show affected area on the map — appropriate visual for spatial air quality data |

### Business Analyst Perspective (Requirements)

| # | Type | Finding |
|---|------|---------|
| 1 | ✅ OK | EU EAQI (1–6) correctly uses EEA official concentration bands. PM2.5: 0/10/20/25/50/75, PM10: 0/20/40/50/100/150 — verified against [EEA reference](https://airindex.eea.europa.eu/AQI/index.html) |
| 2 | ✅ OK | Two alert types with appropriate thresholds: Immediate (EAQI ≥ 5 = Very Poor), Sustained (both halves ≥ 4 = Poor) |
| 3 | ✅ OK | Fetch runs every 15 min (24/7), evaluation runs every 30 min during daytime — good coverage without over-alerting |
| 4 | ✅ OK | `experimental: true` in shared sources — correctly signals this as a beta feature |
| 5 | ⚠️ WARNING | source-trust `0.7` for sensor-community is higher than some municipality sites (which get `0.8`). Citizen science data is inherently noisier — consider if 0.6 would be more appropriate for event matching scoring |
| 6 | ✅ OK | NowCast weighting reduces spike influence — spike AQI < steady AQI verified in tests |

### Developer Perspective (Code Quality)

| # | Type | Finding |
|---|------|---------|
| 1 | ✅ OK | `eaqiFromBands()` correctly uses `fraction * 0.99` to stay within band boundaries — avoids off-by-one at band edges |
| 2 | ✅ OK | `Math.min(..., 6)` cap prevents EAQI exceeding 6 for extreme concentrations |
| 3 | ✅ OK | Outlier filtering: hard cap (999.9 μg/m³) → IQR (1.5× interquartile range) — standard statistical approach |
| 4 | ✅ OK | Sensor count checked AFTER outlier filtering (blocker #3 from v1 review) |
| 5 | ✅ OK | Non-overlapping halves: `[t-4h, t-2h)` and `[t-2h, now]` (blocker #1 fixed) |
| 6 | ✅ OK | Deterministic dedup URL uses 30-min floored timestamp (blocker #2 fixed) |
| 7 | ✅ OK | Terraform resources are conditional on `gcs_readings_bucket != ""` — won't break envs that don't deploy AQ |
| 8 | ✅ OK | `failed` counter + `total: cellReadings.size` in crawler log matches repo convention |
| 9 | 💡 SUGGESTION | `aqi.ts` — the last band `cHigh` values (800 for PM2.5, 1200 for PM10) are arbitrary sentinel values. A comment explaining these are effectively "infinity caps for the interpolation math" would help future readers |
| 10 | ✅ OK | `ReadingsStore` has clean Strategy pattern (GCS vs Local backends) with factory function |
| 11 | ✅ OK | `parseSensorResponse` handles all JS edge cases: string values, null coords, indoor exclusion, NaN filtering |
| 12 | ✅ OK | Pipeline.ts includes `sensor-community` in `EMERGENT_CRAWLERS` array — matches workflow YAML |

### QA Perspective (Testing & Reliability)

| # | Type | Finding |
|---|------|---------|
| 1 | ✅ OK | 21 AQI tests covering: empty input, Good/Poor/Very Poor ranges, PM2.5 vs PM10 max, single-hour, cap at 6, NowCast weighting, label boundaries, category mapping |
| 2 | ✅ OK | 14 parser tests, 9 grid tests, 6 outlier tests, 13 readings store tests — comprehensive unit coverage |
| 3 | ✅ OK | All 1119 tests pass (88 files) |
| 4 | ⚠️ WARNING | No integration test for the full crawler `crawl()` function end-to-end (reading store → grid → outlier → AQI → alert save). Each component is unit-tested, but the composed flow isn't |
| 5 | 💡 SUGGESTION | Edge case: what happens if ALL sensors in a cell are filtered by IQR? Currently `filterOutliers` returns 0 readings → sensor count < 3 → cell skipped. This is correct behavior but an explicit test would document the intention |
| 6 | ✅ OK | `all.yaml` uses sequential `log14` for sensor-community, `emergent.yaml` uses `log4` — both unique within their files |
| 7 | ✅ OK | Terraform conditional resources use `[0]` indexing correctly for IAM binding referencing the bucket |

### Summary

**BLOCKERS: 0** — All previous blockers from v1 review are resolved.

**WARNINGS: 2**
- source-trust score (0.7) may be too high for citizen science data
- No integration test for the full crawler flow

**SUGGESTIONS: 3**
- Add a `deepLinkUrl` pointing to a useful external resource
- Comment the sentinel `cHigh` values in EAQI bands
- Add edge case test for full-IQR-filtered cell

Overall: The implementation is solid, well-tested, and correctly follows the EU EAQI standard. The code is clean, follows repo conventions, and all PR review comments have been addressed. Ready to merge pending the two warnings above (neither is blocking).
